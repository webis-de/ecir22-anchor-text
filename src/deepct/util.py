def clean_qtokens(qtokens):
    ret = {}
    blocklist = load_entrypage_queries()

    for doc, doc_tokens in qtokens.items():
        if len(doc_tokens.keys()) == 0 or len([i for i in doc_tokens.keys() if i in blocklist])/len(doc_tokens.keys()) > 0.2:
            continue
        ret[doc] = doc_tokens

    print('From qtokens with ' + str(len(qtokens)) + ' documents, I filtered terms that may cause train/test overlap so that I now have ' + str(len(ret)) + ' documents.')

    return ret


def load_entrypage_queries():
    from glob import glob
    import importlib
    text_clean = importlib.import_module('train-deep-ct').text_clean

    ret = []
    for f in glob('../../Data/navigational-topics-and-qrels-ms-marco-v1/topics*.tsv'):
        with open(f, 'r') as f:

            for l in f:
                l = l.split('\t')
                if len(l) != 2:
                    raise ValueError('Can not handle ' + str(l))

                ret += text_clean(l[1].strip(), True, True).split(' ')

    return set(ret)


def sample_anchors(anchors_per_doc):
    import json
    import importlib
    text_clean = importlib.import_module('train-deep-ct').text_clean
    blocklist = load_entrypage_queries()
    ret = set()
    for anchors in anchors_per_doc:
        for anchor in set(json.loads(anchors)['anchorText']):
            anchor_terms = text_clean(anchor.strip(), True, True).split(' ')
            if len(anchor_terms) > 0 and len([i for i in anchor_terms if i in blocklist])/len(anchor_terms) < 0.51:
                ret |= set([anchor])

    return ret
