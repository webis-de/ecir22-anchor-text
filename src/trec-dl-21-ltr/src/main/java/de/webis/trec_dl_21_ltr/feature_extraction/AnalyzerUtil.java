package de.webis.trec_dl_21_ltr.feature_extraction;

import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.analysis.ar.ArabicAnalyzer;
import org.apache.lucene.analysis.bn.BengaliAnalyzer;
import org.apache.lucene.analysis.cjk.CJKAnalyzer;
import org.apache.lucene.analysis.core.WhitespaceAnalyzer;
import org.apache.lucene.analysis.da.DanishAnalyzer;
import org.apache.lucene.analysis.de.GermanAnalyzer;
import org.apache.lucene.analysis.es.SpanishAnalyzer;
import org.apache.lucene.analysis.fi.FinnishAnalyzer;
import org.apache.lucene.analysis.fr.FrenchAnalyzer;
import org.apache.lucene.analysis.hi.HindiAnalyzer;
import org.apache.lucene.analysis.hu.HungarianAnalyzer;
import org.apache.lucene.analysis.id.IndonesianAnalyzer;
import org.apache.lucene.analysis.it.ItalianAnalyzer;
import org.apache.lucene.analysis.ja.JapaneseAnalyzer;
import org.apache.lucene.analysis.nl.DutchAnalyzer;
import org.apache.lucene.analysis.no.NorwegianAnalyzer;
import org.apache.lucene.analysis.pt.PortugueseAnalyzer;
import org.apache.lucene.analysis.ru.RussianAnalyzer;
import org.apache.lucene.analysis.sv.SwedishAnalyzer;
import org.apache.lucene.analysis.th.ThaiAnalyzer;
import org.apache.lucene.analysis.tr.TurkishAnalyzer;

import io.anserini.analysis.DefaultEnglishAnalyzer;
import io.anserini.analysis.TweetAnalyzer;
import io.anserini.search.SearchArgs;
import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@UtilityClass
public class AnalyzerUtil {

  private static final Logger LOG = LogManager.getLogger(AnalyzerUtil.class);

  @SneakyThrows
  public static Analyzer analyzer(SearchArgs args) {
	// Copied from SearchCollection#SearchCollection(SearchArgs args)
    if (args.searchtweets) {
      LOG.info("Searching tweets? true");
      return new TweetAnalyzer();
    } else if (args.language.equals("ar")) {
      LOG.info("Language: ar");
      return new ArabicAnalyzer();
    } else if (args.language.equals("bn")) {
      LOG.info("Language: bn");
      return new BengaliAnalyzer();
    } else if (args.language.equals("da")) {
      LOG.info("Language: da");
      return new DanishAnalyzer();
    } else if (args.language.equals("de")) {
      LOG.info("Language: de");
      return new GermanAnalyzer();
    } else if (args.language.equals("es")) {
      LOG.info("Language: es");
      return new SpanishAnalyzer();
    } else if (args.language.equals("fi")) {
      LOG.info("Language: fi");
      return new FinnishAnalyzer();
    } else if (args.language.equals("fr")) {
      LOG.info("Language: fr");
      return new FrenchAnalyzer();
    } else if (args.language.equals("hi")) {
      LOG.info("Language: hi");
      return new HindiAnalyzer();
    } else if (args.language.equals("hu")) {
      LOG.info("Language: hu");
      return new HungarianAnalyzer();
    } else if (args.language.equals("id")) {
      LOG.info("Language: id");
      return new IndonesianAnalyzer();
    } else if (args.language.equals("it")) {
      LOG.info("Language: it");
      return new ItalianAnalyzer();
    } else if (args.language.equals("ja")) {
      LOG.info("Language: ja");
      return new JapaneseAnalyzer();
    } else if (args.language.equals("ko")) {
      LOG.info("Language: ko");
      return new CJKAnalyzer();
    } else if (args.language.equals("nl")) {
      LOG.info("Language: nl");
      return new DutchAnalyzer();
    } else if (args.language.equals("no")) {
      LOG.info("Language: no");
      return new NorwegianAnalyzer();
    } else if (args.language.equals("pt")) {
      LOG.info("Language: pt");
      return new PortugueseAnalyzer();
    } else if (args.language.equals("ru")) {
      LOG.info("Language: ru");
      return new RussianAnalyzer();
    } else if (args.language.equals("sv")) {
      LOG.info("Language: sv");
      return new SwedishAnalyzer();
    } else if (args.language.equals("th")) {
      LOG.info("Language: th");
      return new ThaiAnalyzer();
    } else if (args.language.equals("tr")) {
      LOG.info("Language: tr");
      return new TurkishAnalyzer();
    } else if (args.language.equals("zh")) {
      LOG.info("Language: zh");
      return new CJKAnalyzer();
    } else if (args.pretokenized) {
      LOG.info("Pretokenized");
      return new WhitespaceAnalyzer();
    } else {
      // Default to English
      LOG.info("Language: en");
      LOG.info("Stemmer: " + args.stemmer);
      LOG.info("Keep stopwords? " + args.keepstop);
      LOG.info("Stopwords file " + args.stopwords);
      return DefaultEnglishAnalyzer.fromArguments(args.stemmer, args.keepstop, args.stopwords);
    }
  }
}
