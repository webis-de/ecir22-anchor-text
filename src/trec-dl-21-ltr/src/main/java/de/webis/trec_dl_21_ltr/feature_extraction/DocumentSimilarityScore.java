package de.webis.trec_dl_21_ltr.feature_extraction;

import java.io.IOException;

import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.document.Document;
import org.apache.lucene.index.IndexReader;
import org.apache.lucene.index.Term;
import org.apache.lucene.search.BooleanClause;
import org.apache.lucene.search.BooleanQuery;
import org.apache.lucene.search.IndexSearcher;
import org.apache.lucene.search.Query;
import org.apache.lucene.search.TermQuery;
import org.apache.lucene.search.TopDocs;
import org.apache.lucene.search.similarities.AfterEffectL;
import org.apache.lucene.search.similarities.AxiomaticF2EXP;
import org.apache.lucene.search.similarities.AxiomaticF2LOG;
import org.apache.lucene.search.similarities.BM25Similarity;
import org.apache.lucene.search.similarities.BasicModelIn;
import org.apache.lucene.search.similarities.ClassicSimilarity;
import org.apache.lucene.search.similarities.DFRSimilarity;
import org.apache.lucene.search.similarities.DistributionSPL;
import org.apache.lucene.search.similarities.IBSimilarity;
import org.apache.lucene.search.similarities.LMDirichletSimilarity;
import org.apache.lucene.search.similarities.LMJelinekMercerSimilarity;
import org.apache.lucene.search.similarities.LambdaDF;
import org.apache.lucene.search.similarities.NormalizationH2;
import org.apache.lucene.search.similarities.Similarity;

import io.anserini.index.IndexArgs;
import io.anserini.search.SearchArgs;
import io.anserini.search.query.BagOfWordsQueryGenerator;

/**
 * Feature extraction for TREC DL 2021.
 * 
 * @author Maik Fr√∂be
 *
 */
public class DocumentSimilarityScore {

  private final IndexReader indexReader;
  
  private final SearchArgs args;
  
  private final Analyzer analyzer;
  
  public DocumentSimilarityScore(IndexReader indexReader) {
    this.indexReader = indexReader;
    this.args = new SearchArgs();
    this.analyzer = AnalyzerUtil.analyzer(this.args);
  }

  public float tfSimilarity(String query, String documentId) {
    return calculate(new TfSimilarity(), query, documentId);
  }
  
  public float tfIdfSimilarity(String query, String documentId) {
    return calculate(new ClassicSimilarity(), query, documentId);
  }

  public float bm25Similarity(String query, String documentId) {
    return calculate(bm25(), query, documentId);
  }

  private BM25Similarity bm25() {
    return new BM25Similarity(uniqueFloat(args.bm25_k1), uniqueFloat(args.bm25_b));
  }

  public float pl2Similarity(String query, String documentId) {
    return calculate(pl2(), query, documentId);
  }
  
  private DFRSimilarity pl2() {
    return new DFRSimilarity(new BasicModelIn(), new AfterEffectL(), new NormalizationH2(uniqueFloat(args.inl2_c)));
  }

  public float qlSimilarity(String query, String documentId) {
    return calculate(ql(), query, documentId);
  }

  private LMDirichletSimilarity ql() {
    return new LMDirichletSimilarity(uniqueFloat(args.qld_mu));
  }
  
  public float qljmSimilarity(String query, String documentId) {
    return calculate(qljm(), query, documentId);
  }
  
  private LMJelinekMercerSimilarity qljm() {
	  return new LMJelinekMercerSimilarity(uniqueFloat(args.qljm_lambda));
  }
  
  public float splSimilarity(String query, String documentId) {
    return calculate(spl(), query, documentId);
  }
  
  private IBSimilarity spl() {
	 return new IBSimilarity(new DistributionSPL(), new LambdaDF(),  new NormalizationH2(uniqueFloat(args.spl_c)));
  }
  
  public float f2expSimilarity(String query, String documentId) {
    return calculate(f2exp(), query, documentId);
  }
  
  private AxiomaticF2EXP f2exp() {
    return new AxiomaticF2EXP(uniqueFloat(args.f2exp_s));
  }
  
  public float f2logSimilarity(String query, String documentId) {
    return calculate(f2log(), query, documentId);
  }

  private AxiomaticF2LOG f2log() {
    return new AxiomaticF2LOG(uniqueFloat(args.f2log_s));
  }
  
  public float calculate(Similarity similarity, String query, String documentId) {
    return calculate(searcher(similarity, indexReader), query, documentId);
  }
  
  public float calculate(IndexSearcher searcher, String query, String documentId) {
    return calculate(searcher, similarityInBody(query), documentId);
  }

  public static float calculate(IndexSearcher searcher, Query query, String documentId) {
    BooleanQuery scoreForDoc = new BooleanQuery.Builder()
      .add(idIs(documentId), BooleanClause.Occur.FILTER)
      .add(query, BooleanClause.Occur.MUST)
      .build();

    return retrieveScoreOrFail(searcher, scoreForDoc, documentId);
  }

  private static IndexSearcher searcher(Similarity similarity, IndexReader reader) {
    IndexSearcher ret = new IndexSearcher(reader);
    ret.setSimilarity(similarity);
    
    return ret;
  }

  private static float retrieveScoreOrFail(IndexSearcher searcher, Query query, String docId) {
    try {
      return retrieveScore(searcher, query, docId);
    } catch(IOException e) {
      throw new RuntimeException(e);
    }
  }
  
  private static float retrieveScore(IndexSearcher searcher, Query query, String docId) throws IOException {
    TopDocs ret = searcher.search(query, 10);

    if(ret.scoreDocs.length == 0) {
      return 0;
    } else if (ret.scoreDocs.length == 1) {
      return ret.scoreDocs[0].score;
    }
    
    Document firstDoc = searcher.getIndexReader().document(ret.scoreDocs[0].doc);
    String actualId = firstDoc.get(IndexArgs.ID);

    if(docId == null || !docId.equals(actualId)) {
      throw new RuntimeException("I expected a document with id '" + docId + "', but got '" + actualId + "'.");
    }

    return ret.scoreDocs[0].score;
  }
  
  private static Query idIs(String documentId) {
    return new TermQuery(new Term(IndexArgs.ID, documentId));
  }

  private Query similarityInBody(String queryString) {
    return new BagOfWordsQueryGenerator().buildQuery(IndexArgs.CONTENTS, analyzer, queryString);
  }
  
  private static float uniqueFloat(String[] choices) {
    return Float.valueOf(uniqueElementOfFail(choices));
  }
  
  private static String uniqueElementOfFail(String[] choices) {
    if(choices.length != 1) {
      throw new RuntimeException("Cant extract unique element from array of length " + choices.length);
    }
    
    return choices[0];
  }
}