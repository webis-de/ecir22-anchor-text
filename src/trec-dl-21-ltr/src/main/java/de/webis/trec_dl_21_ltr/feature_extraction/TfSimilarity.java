package de.webis.trec_dl_21_ltr.feature_extraction;

import org.apache.lucene.search.similarities.ClassicSimilarity;

/**
 * TF-Similarity as feature for TREC-DL 2021.
 * 
 * @author Maik Fr√∂be
 *
 */
public class TfSimilarity extends ClassicSimilarity {
  @Override
  public float idf(long docFreq, long docCount) {
    return 1.0f;
  }
  
  @Override
  public float lengthNorm(int numTerms) {
    return 1.0f;
  }
  
  @Override
  public float tf(float freq) {
    return freq;
  }
}