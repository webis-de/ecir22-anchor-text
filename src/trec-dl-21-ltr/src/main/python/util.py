def write_run_file(model_file, feature_file, system_name, out_file):
    with open(out_file, 'w') as f:
        f.write(run_file_content(model_file, feature_file, system_name))

def run_file_content(model_file, feature_file, system_name):
    from lightgbm import Booster
    model = Booster(model_file=model_file)
    ret = ''

    to_rerank = load_ranking_data(feature_file)
    for topic_to_rerank in to_rerank:
        doc_to_score = []
        y_predictions = model.predict(topic_to_rerank['documentFeatures'])
        assert len(y_predictions) == len(topic_to_rerank['documents'])

        for i in range(len(topic_to_rerank['documents'])):
            doc_to_score += [{'documentId': topic_to_rerank['documents'][i], 'score': y_predictions[i]}]

        doc_to_score = sorted(doc_to_score, key=lambda i: i['score'], reverse=True)
        ret += topic_as_run_file(topic_to_rerank['topic'], doc_to_score, system_name)

    return ret.strip()

def topic_as_run_file(topic_nr, doc_to_score, system_name):
    ret = ''

    for i in range(len(doc_to_score)):
        ret += topic_nr + ' Q0 ' + doc_to_score[i]['documentId'] + ' ' + str(i+1) + ' ' + str(doc_to_score[i]['score']) + ' ' + system_name + '\n'

    return ret

def load_ranking_data(file_name):
    import json
    ret = []
    with open(file_name, 'r') as f:
        for l in f:
            l = json.loads(l)

            ret += [{
                'topic': l['topic'],
                'documents': [i['documentId'] for i in l['documents']],
                'documentFeatures': as_np_array(l['documents']),
            }]
    return ret


def as_np_array(docs):
    import numpy as np
    ret = []
    for doc in docs:
        last_feature=1
        features = doc['features']
        # relevance label and query id
        ret_features = [-100, -100]
        for feature_num in features:
            if (int(feature_num) -1) != last_feature:
                raise ValueError('Could not handle...')

            ret_features += [features[feature_num]]
            last_feature = int(feature_num)
        ret += [ret_features]

    return np.array(ret)

