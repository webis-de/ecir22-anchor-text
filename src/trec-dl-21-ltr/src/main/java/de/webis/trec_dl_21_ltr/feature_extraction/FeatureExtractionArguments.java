package de.webis.trec_dl_21_ltr.feature_extraction;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.LinkedHashMap;
import java.util.Map;

import de.webis.trec_dl_21_ltr.parsing.QueryExtractor;
import lombok.Builder;
import lombok.Data;

@Data
@Builder
public class FeatureExtractionArguments {
  private final String indexDir, outputDir, runFileDir;

  private final Map<String, String> fieldToIndex;
  
  private QueryExtractor queryExtractor;

  public Path outputFileForTopic(int topic) {
    return Paths.get(outputDir).resolve(topic + ".json");
  }

  public static Map<String, String> msMarcoV2Features() {
    Map<String, String> ret = new LinkedHashMap<>();
    ret.put("body", "msmarco-doc-v2");
    ret.put("orcas", "orcas-ms-marco-v2");
    ret.put("anchor-comb", "cc-union-16-to-21-anchortext-1000-v2");
    ret.put("url", "msmarco-document-v2-url-only.pos+docvectors+raw");
    ret.put("title", "msmarco-document-v2-title-only.pos+docvectors+raw");
    
    return ret;
  }
  
  public static Map<String, String> msMarcoV2FeaturesTrecDL21() {
    Map<String, String> ret = new LinkedHashMap<>();
    ret.put("body", "msmarco-doc-v2");
    ret.put("anchor", "cc-19-47-anchortext-v2");
    ret.put("url", "msmarco-document-v2-url-only.pos+docvectors+raw");
    ret.put("title", "msmarco-document-v2-title-only.pos+docvectors+raw");

    return ret;
  }
  
  public static Map<String, String> msMarcoV1Features() {
    Map<String, String> ret = new LinkedHashMap<>();
    ret.put("body", "ms-marco-content"); 
    ret.put("url", "msmarco-document-v1-url-only.pos+docvectors+raw");
    ret.put("title", "msmarco-document-v1-title-only.pos+docvectors+raw");
    ret.put("anchor-16-07", "cc-16-07-anchortext");
    ret.put("anchor-17-04", "cc-17-04-anchortext");
    ret.put("anchor-18-13", "cc-18-13-anchortext");
    ret.put("anchor-19-47", "cc-19-47-anchortext");
    ret.put("anchor-20-05", "cc-20-05-anchortext");
    ret.put("anchor-21-04", "cc-21-04-anchortext");
    ret.put("anchor-comb", "cc-combined-anchortext");
    ret.put("anchor-comb-no-16", "cc-combined-no16-anchortext");
    ret.put("orcas", "orcas");
    
    return ret;
  }
}
