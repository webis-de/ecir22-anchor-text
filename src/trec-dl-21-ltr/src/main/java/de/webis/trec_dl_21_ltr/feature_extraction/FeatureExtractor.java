package de.webis.trec_dl_21_ltr.feature_extraction;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import com.fasterxml.jackson.databind.ObjectMapper;

import lombok.Data;
import lombok.SneakyThrows;

@Data
public class FeatureExtractor {
  private final Map<Integer, String> idToQuery;
  private final Map<Integer, Set<String>> topicToDocs;

  private static final Logger LOG = LogManager.getLogger(FeatureExtractor.class);
  
  public void runFeatureExtraction(FeatureExtractionArguments args) {
    topicToDocs.entrySet().parallelStream().forEach(i -> runFeatureExtraction(i.getKey(), i.getValue(), args));
  }

  @SneakyThrows
  private void runFeatureExtraction(Integer topic, Set<String> docs, FeatureExtractionArguments args) {
    Path outputFile = args.outputFileForTopic(topic);
    if(outputFile.toFile().exists()) {
      LOG.info("Skip existing output: " + topic);
      return;
    }
    
    Files.writeString(outputFile, calculateFeatures(topic, docs, args));
  }

  @SneakyThrows
  public String calculateFeatures(Integer topic, Set<String> docs, FeatureExtractionArguments args) {
    long start = System.currentTimeMillis();

    String query = idToQuery.get(topic);
    if (query == null || query.trim().isEmpty()) {
      throw new RuntimeException("Could not find query for topic " + topic);
    }

    LOG.info("Start extracting features for topic " + topic + " with " + docs.size() + " documents.");
    SimilarityCalculator calc = new SimilarityCalculator(args);
    Map<String, Map<String, Float>> features = calc.calculateFeaturesForDocuments(query, new ArrayList<>(docs));
    
    Map<String, Object> ret = new LinkedHashMap<>();
    ret.put("query", query);
    ret.put("topic", topic);
    ret.put("documentFeatures", features);
    
    LOG.info("Finished extraction of features for topic " + topic + " with " + docs.size() + " documents in " + (System.currentTimeMillis() - start) + "ms.");

    return new ObjectMapper().writeValueAsString(ret);
  }
}
