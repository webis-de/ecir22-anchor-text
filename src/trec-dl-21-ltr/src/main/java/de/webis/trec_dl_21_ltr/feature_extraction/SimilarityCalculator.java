package de.webis.trec_dl_21_ltr.feature_extraction;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.apache.lucene.index.DirectoryReader;
import org.apache.lucene.index.IndexReader;
import org.apache.lucene.store.FSDirectory;

import lombok.SneakyThrows;

public class SimilarityCalculator {

  private final Map<String, DocumentSimilarityScore> docSimilarityScores;

  private final FeatureExtractionArguments args;
  
  public SimilarityCalculator(FeatureExtractionArguments args) {
    this.docSimilarityScores = initializeSimilarityReaders(args);
    this.args = args;
  }

  public Map<String, Map<String, Float>> calculateFeaturesForDocuments(String query, List<String> documentIds) {
    Map<String, Map<String, Map<String, Float>>> fieldToDocScores = new LinkedHashMap<>();
    
    for(String field: args.getFieldToIndex().keySet().stream().sorted().collect(Collectors.toList())) {
      fieldToDocScores.put(field, calculateFeaturesForDocumentsOnField(query, documentIds, field));
    }
    
    Map<String, Map<String, Float>> ret = new LinkedHashMap<>();
    for(String doc: documentIds.stream().sorted().collect(Collectors.toList())) {
      Map<String, Float> docFeatures = new LinkedHashMap<>();
      for(String field: args.getFieldToIndex().keySet()) {
        Map<String, Float> tmp = fieldToDocScores.get(field).get(doc);
        if(tmp == null || tmp.isEmpty()) {
          throw new RuntimeException("This should not happen.");
        }
        
        docFeatures.putAll(tmp);
      }
      
      ret.put(doc, docFeatures);
    }

    return ret;
  }

  private Map<String, Map<String, Float>> calculateFeaturesForDocumentsOnField(String query, List<String> documentIds, String field) {
    Map<String, Map<String, Float>> ret = new LinkedHashMap<>();

    for(String docId: documentIds) {
      ret.put(docId, calculateFeaturesForDocumentOnField(query, docId, field));
    }

    return ret;
  }
  
  @SneakyThrows
  private Map<String, Float> calculateFeaturesForDocumentOnField(String query, String documentId, String field) {
    DocumentSimilarityScore sim = docSimilarityScores.get(field);
    Map<String, Float> ret = new LinkedHashMap<>();
    ret.put(field + "-bm25-similarity", sim.bm25Similarity(query, documentId));
    ret.put(field + "-f2exp-similarity", sim.f2expSimilarity(query, documentId));
    ret.put(field + "-spl-similarity", sim.splSimilarity(query, documentId));
    ret.put(field + "-pl2-similarity", sim.pl2Similarity(query, documentId));
    ret.put(field + "-ql-similarity", sim.qlSimilarity(query, documentId));
    ret.put(field + "-qljm-similarity", sim.qljmSimilarity(query, documentId));
    ret.put(field + "-f2log-similarity", sim.f2logSimilarity(query, documentId));
    ret.put(field + "-tf-idf-similarity", sim.tfIdfSimilarity(query, documentId));
    ret.put(field + "-tf-similarity", sim.tfSimilarity(query, documentId));
    
    return ret;
  }

  @SneakyThrows
  private static Map<String, DocumentSimilarityScore> initializeSimilarityReaders(FeatureExtractionArguments args) {
    Map<String, DocumentSimilarityScore> ret = new LinkedHashMap<>();

    for(String field : args.getFieldToIndex().keySet()) {
      IndexReader reader = DirectoryReader.open(FSDirectory.open(indexPathForField(args, field)));
      ret.put(field, new DocumentSimilarityScore(reader));
    }

    return ret;
  }

  private static Path indexPathForField(FeatureExtractionArguments args, String field) {
    return Paths.get(args.getIndexDir()).resolve(args.getFieldToIndex().get(field));
  }
}