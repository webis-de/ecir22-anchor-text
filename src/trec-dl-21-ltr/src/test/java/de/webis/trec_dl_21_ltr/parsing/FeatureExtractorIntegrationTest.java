package de.webis.trec_dl_21_ltr.parsing;

import java.util.Map;
import java.util.Set;

import org.approvaltests.Approvals;
import org.junit.Test;

import de.webis.trec_dl_21_ltr.feature_extraction.FeatureExtractionArguments;
import de.webis.trec_dl_21_ltr.feature_extraction.FeatureExtractor;

public class FeatureExtractorIntegrationTest {
  
  @Test
  public void approveWithQueryFromTitleIndex() {
    //{"id":"msmarco_doc_00_4840315","contents":"Plants and Animals of the Everglades"}
    FeatureExtractor extractor = new FeatureExtractor(
      Map.of(-1, "Plants and Animals of the Everglades"),
      Map.of(-1, Set.of("msmarco_doc_00_4840315"))
    );

    Approvals.verify(extractor.calculateFeatures(-1, Set.of("msmarco_doc_00_4840315"), msMarcoV2Args()));
  }
  
  @Test
  public void approveV1DocWithQueryFromTitleIndex() {
    //{"id":"D1885729","contents":"How to Kill Weeds Without Killing Plants"}
    FeatureExtractor extractor = new FeatureExtractor(
      Map.of(-1, "How to Kill Weeds Without Killing Plants"),
      Map.of(-1, Set.of("D1885729"))
    );

    Approvals.verify(extractor.calculateFeatures(-1, Set.of("D1885729"), msMarcoV1Args()));
  }
  
  @Test
  public void approveWithQueryFromUrlIndex() {
    //{"id":"msmarco_doc_00_178958616","contents":"http://airportchandigarh.com/transport/chandigarh-volvo-bus-schedules"}
    FeatureExtractor extractor = new FeatureExtractor(
      Map.of(-1, "http://airportchandigarh.com/transport/chandigarh-volvo-bus-schedules"),
      Map.of(-1, Set.of("msmarco_doc_00_178958616"))
    );

    Approvals.verify(extractor.calculateFeatures(-1, Set.of("msmarco_doc_00_178958616"), msMarcoV2Args()));
  }
  
  @Test
  public void approveWithQueryFromAnchorIndex() {
    //{"id":"msmarco_doc_17_2981346117","contents":"Join Together (song) The Who reference Join Together"}
    FeatureExtractor extractor = new FeatureExtractor(
      Map.of(-1, "Join Together (song) The Who reference Join Together"),
      Map.of(-1, Set.of("msmarco_doc_17_2981346117"))
    );

    Approvals.verify(extractor.calculateFeatures(-1, Set.of("msmarco_doc_17_2981346117"), msMarcoV2Args()));
  }
  
  @Test
  public void approveWithV1DocQueryFromAnchorIndex() {
    FeatureExtractor extractor = new FeatureExtractor(
      Map.of(-1, "Erstellt mit Storefront & WooCommerce"),
      Map.of(-1, Set.of("D2909755"))
    );

    Approvals.verify(extractor.calculateFeatures(-1, Set.of("D2909755"), msMarcoV1Args()));
  }
  
  @Test
  public void approveWithQueryFromRunFile() {
    // 257309 Q0 msmarco_doc_01_1995043585 1 15.637100 Anserini
    // 257309 how long does it take to get your bsrn if you already have a bachelors degree
    FeatureExtractor extractor = new FeatureExtractor(
      Map.of(-1, "how long does it take to get your bsrn if you already have a bachelors degree"),
      Map.of(-1, Set.of("msmarco_doc_17_2981346117"))
    );

    Approvals.verify(extractor.calculateFeatures(-1, Set.of("msmarco_doc_17_2981346117"), msMarcoV2Args()));
  }
  
  private static FeatureExtractionArguments msMarcoV1Args() {
    return FeatureExtractionArguments.builder()
      .indexDir("/anserini-indexes")
      .fieldToIndex(FeatureExtractionArguments.msMarcoV1Features())
      .build();
  }
  
  private static FeatureExtractionArguments msMarcoV2Args() {
    return FeatureExtractionArguments.builder()
      .indexDir("/anserini-indexes")
      .fieldToIndex(FeatureExtractionArguments.msMarcoV2Features())
      .build();
  }
}
