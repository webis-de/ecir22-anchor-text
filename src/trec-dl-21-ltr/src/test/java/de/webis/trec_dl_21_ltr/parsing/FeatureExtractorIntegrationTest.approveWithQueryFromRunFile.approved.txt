{"query":"how long does it take to get your bsrn if you already have a bachelors degree","topic":-1,"documentFeatures":{"msmarco_doc_17_2981346117":{"body-bm25-similarity":2.495199,"body-f2exp-similarity":5.1353216,"body-spl-similarity":2.8710442,"body-pl2-similarity":1.4150404,"body-ql-similarity":1.5612885,"body-qljm-similarity":12.75075,"body-f2log-similarity":2.4181309,"body-tf-idf-similarity":0.4831325,"body-tf-similarity":29.0,"anchor-bm25-similarity":0.0,"anchor-f2exp-similarity":0.0,"anchor-spl-similarity":0.0,"anchor-pl2-similarity":0.0,"anchor-ql-similarity":0.0,"anchor-qljm-similarity":0.0,"anchor-f2log-similarity":0.0,"anchor-tf-idf-similarity":0.0,"anchor-tf-similarity":0.0,"url-bm25-similarity":0.0,"url-f2exp-similarity":0.0,"url-spl-similarity":0.0,"url-pl2-similarity":0.0,"url-ql-similarity":0.0,"url-qljm-similarity":0.0,"url-f2log-similarity":0.0,"url-tf-idf-similarity":0.0,"url-tf-similarity":0.0,"title-bm25-similarity":0.0,"title-f2exp-similarity":0.0,"title-spl-similarity":0.0,"title-pl2-similarity":0.0,"title-ql-similarity":0.0,"title-qljm-similarity":0.0,"title-f2log-similarity":0.0,"title-tf-idf-similarity":0.0,"title-tf-similarity":0.0}}}